merlin/json.py

```diff
@@ -0,0 +1,177 @@
+        """
+        # print(lvl * "  " + str(type(obj)))
+        if obj is None or type(obj) in {str, int, float, bool}:
+            return obj
+        if type(obj) == list:
+            result = []
+            for item in obj:
+                result.append(MerlinEncoder.to_dict(item, lvl + 1))
+            return {"__list__": result}
+        if type(obj) == tuple:
+            result = []
+            for item in obj:
+                result.append(MerlinEncoder.to_dict(item, lvl + 1))
+            return {"__tuple__": tuple(result)}
+        if type(obj) == dict:
+            result = {}
+            for k, v in obj.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return result
+        if isinstance(obj, enum.Enum):
+            return {"__enum__": str(obj)}
+        if isinstance(obj, np.ndarray):
+            return {"__ndarray__": obj.tolist()}
+        if isinstance(obj, Variable):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__Variable__": result}
+        if isinstance(obj, StudyStep):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__StudyStep__": result}
+        if isinstance(obj, set):
+            result = []
+            for item in obj:
+                result.append(MerlinEncoder.to_dict(item, lvl + 1))
+            return {"__set__": result}
+        if isinstance(obj, deque):
+            result = []
+            for item in obj:
+                result.append(MerlinEncoder.to_dict(item, lvl + 1))
+            return {"__deque__": result}
+        if isinstance(obj, OrderedDict):
+            result = {}
+            for k, v in obj.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__OrderedDict__": result}
+        if isinstance(obj, _StepRecord):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"___StepRecord__": result}
+        if isinstance(obj, ExecutionGraph):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__ExecutionGraph__": result}
+        if isinstance(obj, MerlinDAG):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__MerlinDAG__": result}
+        if isinstance(obj, MerlinStep):
+            result = {}
+            for k, v in obj.__dict__.items():
+                result[k] = MerlinEncoder.to_dict(v, lvl + 1)
+            return {"__MerlinStep__": result}
+        else:
+            raise TypeError(f"Type '{type(obj)}' not supported by MerlinEncoder!")
+
+    def encode(self, obj):
```

Comment:
I think this should have a try for the to_dict.
```
        try:
            d = MerlinEncoder.to_dict(obj)
            return json.dumps(d, indent=4)
        except TypeError:
            return json.dumps(d, indent=4)
        except Exception as e:
            print(f"Broke on object of type {type(obj)}: {obj}")
            raise e
```
   and then a similar block for the decode.

Reply:
agreed
a try - catch seems cleaner
should the default be for regular json then catch the typeError? (the above message throws a kombu EncodeError)

Reply:
We should certainly try it with the logic reversed, (json only first). Ben did you already try that?

Reply:
I added the TypeError except and am now looking at reversing the logic. 

By my tests, it is already able to successfully serialize then deserialize all these types: list, tuple, enum, ndarray, Variable, StudyStep, set, deque, OrderedDict, _StepRecord, ExecutionGraph, MerlinDAG, MerlinStep